src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.MSBuild.Build;
+using Roslyn.Utilities;
+using MSB = Microsoft.Build;
+
+namespace Microsoft.CodeAnalysis.MSBuild
+{
+    /// <summary>
+    /// An API for loading msbuild project files.
+    /// </summary>
+    public partial class MSBuildProjectLoader
+    {
+        // the workspace that the projects and solutions are intended to be loaded into.
+        private readonly Workspace _workspace;
+
+        private readonly DiagnosticReporter _diagnosticReporter;
+        private readonly PathResolver _pathResolver;
+        private readonly ProjectFileLoaderRegistry _projectFileLoaderRegistry;
+
+        // used to protect access to the following mutable state
+        private readonly NonReentrantLock _dataGuard = new NonReentrantLock();
+        private ImmutableDictionary<string, string> _properties;
+
+        internal readonly ProjectBuildManager BuildManager;
+
+        internal MSBuildProjectLoader(
+            Workspace workspace,
+            DiagnosticReporter diagnosticReporter,
+            ProjectFileLoaderRegistry projectFileLoaderRegistry,
+            ImmutableDictionary<string, string> properties)
+        {
+            _workspace = workspace;
+            _diagnosticReporter = diagnosticReporter ?? new DiagnosticReporter(workspace);
+            _pathResolver = new PathResolver(_diagnosticReporter);
+            _projectFileLoaderRegistry = projectFileLoaderRegistry ?? new ProjectFileLoaderRegistry(workspace, _diagnosticReporter);
+            _properties = properties ?? ImmutableDictionary<string, string>.Empty;
```

Comment:
>ImmutableDictionary<string, string>.Empty; [](start = 40, length = 42)

Should this be specifying a case insensitive comparer?

Reply:
Probably, yes. Otherwise, there's a bug if the user passes us a dictionary with `solutiondir` already set.